src/main/java/org/skriptlang/skript/lang/condition/Conditional.java

```diff
@@ -0,0 +1,253 @@
+package org.skriptlang.skript.lang.condition;
+
+import ch.njol.skript.lang.Debuggable;
+import ch.njol.util.Kleenean;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An object which can evaluate to `true`, `false`, or `unknown`.
+ * `unknown` is currently unused, but intended for future handling of unexpected runtime situations, where some aspect of
+ * the condition in ill-defined by the user and would result in ambiguous or undefined behavior.
+ * @param <T> The context class to use for evaluation.
+ */
+// TODO: replace Bukkit event with proper context object
+public interface Conditional<T> extends Debuggable {
+
+	/**
+	 * Evaluates this object as `true`, `false`, or `unknown`.
+	 * This value may change between subsequent callings.
+	 *
+	 * @param context The context with which to evaluate this object.
+	 * @return The evaluation of this object.
+	 */
+	@Contract(pure = true)
+	Kleenean evaluate(T context);
+
+	/**
+	 * Evaluates this object as `true`, `false`, or `unknown`.
+	 * This value may change between subsequent callings.
+	 * May use a mutable cache of evaluated conditionals to prevent duplicate evaluations.
+	 *
+	 * @param context The context with which to evaluate this object.
+	 * @param cache The cache of evaluated conditionals.
+	 * @return The evaluation of this object.
+	 */
+	@Contract(pure = true)
+	default Kleenean evaluate(T context, @Nullable Map<Conditional<T>, Kleenean> cache) {
+		if (cache == null)
+			return evaluate(context);
+		//noinspection DataFlowIssue
+		return cache.computeIfAbsent(this, (cond -> cond.evaluate(context)));
```

Comment:
```suggestion
		return cache.computeIfAbsent(this, cond -> cond.evaluate(context));
``` 
if they're required feel free to ignore